### ISSUE TITLE
New webpage for Pester

### ISSUE THREAD

**Issue by nohwnd** [pester/pester.github.com#7]
Added people to this repo so they can talk and share ideas about how to go forward with this. What I have in mind is a web built on some static page generator like Jekyll or Hexo or similar. This would allow people to contribute easily via PRs in a readable manner with minimal requirements to learn new skills. 

There should be few major things: 
- main page with general info about Pester
- documentation that will be generated from markdown and used to replace wiki and will integrate with comment based help
- articles, where we can publish articles, tips and and examples and where links to other blogs can be shared

- ideally the page would look simple and modern (I am looking at https://dbatools.io/ and like it a lot)

I can help with some of the stuff, except for design, I am terrible at that, so ping me if anything is needed 🙂

**Comment by nohwnd** [comment]
@scrthq I really like the designs you've shown, would you take leadership on the design part of this please? 🙏

**Comment by nohwnd** [comment]
@bravo-kernel docosaurus and docosaurus.powershell look great, I am not sure how much we can cusomize the design, but it looks like a great thing to integrate.

**Comment by vexx32** [comment]
I am most familiar with Jekyll, but tbh once the framework's in place I'll be able to figure it out.

@scrthq if you want to take lead I'm all for it. If not, I can grab a nice-looking Jekyll template that we can use as a starting point and modify as we go? 😊 

Just let me know what y'all need, happy to help! 💖

**Comment by MarkWarneke** [comment]
Can you share the design from @scrthq @nohwnd which you mentioned?
Are we going to build the Jekyll stuff from scratch?

I have a nice template that I used and thus I am quite familiar with - it is called [Beautiful-jekyll](https://github.com/daattali/beautiful-jekyll#readme).
The HTMl and CSS can be fully customized.

**Comment by MarkWarneke** [comment]
We could use [HTML5UP](https://html5up.net/) as a design reference too.

**Comment by bravo-kernel** [comment]
Friendly people, docusaurus is fully theme-able. I am in the middle of an exam-track but will create a new ticket with all required steps as soon as I am done with that. I will create a task-list for the issues to make clear how this should work and what steps are required. @nohwnd can then make the decision to use it or not.

A little more patience please, it will make sense very soon.

**Comment by bravo-kernel** [comment]
https://docusaurus-powershell.netlify.com/

**Comment by gaelcolas** [comment]
Just wanted to say that you should add the videos to pester talks (your YT playlist @nohwnd ) so it's easily discoverable.

At DSCcommunity.org we use Hugo (and we aren't great about design, but content is what matters).

**Comment by adilio** [comment]
Hi All,

Great discussions so far! I feel like we should come to a consensus in this thread as to what tooling we are using going forward, as that will largely drive the later steps.

Jakub (@nohwnd) and Joel (@vexx32), I agree that Nate's (@scrthq) work on other PowerShell project would make him a good fit to lead this charge.

As reference from the question above, examples of work that Nate has done on his own PowerShell projects sites is here:

Jekyll based: 
https://vaporshell.io (HTML5UP)
https://ferrell.io

MkDocs based: 
https://psgsuite.io 

Whatever process we choose, it needs to be wicked easy to add and update content (with some automation), as @gaelcolas mentioned that is the most important.

Hope that makes sense?

**Comment by vexx32** [comment]
I like the VaporShell look as a starting point! 👍

**Comment by scrthq** [comment]
hey y'all - sorry for the lag here! putting out fires at work.

Very happy to help where I can! The code is up on the GitHub repos, with the VaporShell page being on the github-pages branch of that repo (https://github.com/scrthq/VaporShell/tree/gh-pages)

VaporShell is using a free theme for Jekyll that I've modified quite a bit (adjusted CSS to make it feel more like "One Dark Pro", etc). Feel free to fork and adjust as a base point!

My blog (https://ferrell.io/) uses a paid version of the HydeJack theme for Jekyll. Happy to share what I can, but I don't think that layout is fitting for this type of project anyway IMHO.

MkDocs is pretty flexible but I'm not as well versed as Jekyll since it's pretty new to me.



All of the above sites are built on `git push`, effectively

**Comment by nohwnd** [comment]
@MarkWarneke I was talking about this tweet, same pages are linked above. https://twitter.com/scrthq/status/1206964817979957248

**Comment by nohwnd** [comment]
Looks like setting up another repo to just manage the docs would give nice separation of privilege and would also allow for docs.pester.dev to be published. That would then allow to use docusaurus easily without really integrating it with the jekyll hosted site (not sure if it would be a problem, just thinking out loud here). Then the theme on both pester.dev and docs.pester.dev can be made similar to make it feel a bit like one site.

**Comment by bravo-kernel** [comment]
I have created a list of features that might be used to consider which solution to go for: https://github.com/pester/pester.github.com/issues/8

**Comment by glennsarti** [comment]
Thoughts:

I use Jekyll for my personal blog and Hugo for the puppet-vscode docs site

* I like the https://vaporshell.io (We can bikeshed font choice later 😁)

Copy-paste and make edits.

Note - This is licensed under Attribution 3.0 Unported (CC BY 3.0) so this project would need to do the same if modifying that work ... which brings me to

* Need to decide how the website content will be licensed.

* Both jeykyll and hugo have excellent Netlify support for both Per-PR and auto deploy scenarios

* From a contributor point of view, Jeykyll is (much) slower than Hugo (Ruby vs Go).  Getting people to install Ruby for jekyll (particularly on Windows) is a major chore (I say this is a full-time ruby dev on Windows). Also jekyll dropped live reload support on Windows.

So if you want inexperienced people, on Windows, to write and test doc changes locally, I would avoid Jekyll in preference for Hugo

IMO, you want to avoid this kind of thing for get people onboarded:

``` text
To contribute to the Pester documentation, just do these simple steps:
1. Install WSL
2. Install ruby in WSL
3. gem update --system (accept all prompts)
4. bundle install --path <something.  (Ignore the warnings about deprecation that just started this week)
5. bundle exec jekyll serve ...
```

This is a big reason why I like the docsy.dev Hugo template. It has native support for adding buttons to get people to edit docs, FROM the docs.
![image](https://user-images.githubusercontent.com/5819622/71141424-0f240100-224f-11ea-8183-69fa3d563cfb.png)

* Avoid SPAs. Sure they're pretty, but they (normally) break deep-linking.  I would fully expect that the Help url for all of the pester command-lets would eventually lead to this new website, so you need deep linking for that. e.g.

``` powershell
> Get-Help Invoke-Pester -Online

<opens browser to https://pester.dev/documentation/latest/Invoke-Pester>
```

* (My opinion) Right _now_, avoid doing crazy stuff. Keep it simple, and just plain, get something out there!  Markdown plus a pretty theme will already be a much better experience.

You can be crazy later. I changed entire blog engines (Jekyll -> Hugo) for the puppet-vscode docs site. Users don't care (authors do)

**Comment by glennsarti** [comment]
Oh and, is the intention to support/host multiple versions of the Pester documentation or only one?

If one, would that be latest unreleased, latest released?

**Comment by nohwnd** [comment]
Just a note: I wanted to try it, and made another repo Pester/docs and it is published to https://docs.pester.dev It depends on the gh-pages branch, so imho I could even keep it in the pester/pester repo and automate the updates when docs folder changes or something. That would allow the docs to be close to the actual code and version with it. But for the rest of the web I want to avoid mixing it with the main repo, mainly because monitoring issues and PRs is easier that way.

**Comment by glennsarti** [comment]
@nohwnd Yup saw that. We can bike-shed the gh-pages branch thing later.

I'm all for forking vaporshell as a place to start as long as you're happy with the license implication.  While I'm fairly sure @scrthq doesn't mind (😁) I need to still be seen as lawful-good and respect the licensing.

**Comment by bravo-kernel** [comment]
@glennsarti:

- do you mean deeplinks like this https://docusaurus-powershell.netlify.com/docs/cmdlets/new-docusaurushelp/
- the reasoning to "avoid SPAs" does not apply to Docusaurus. Remember that this is facebook's primary documentation tool used for almost all their software. Most if not all paint-points have already been tackled by those friends, 2020 🎉

```
To contribute to the Pester documentation, just do these (non-WSL) simple steps:
yarn install
yarn develop
```

_What will be more present on people's computers nowadays? Ruby or nodejs_

**Comment by bravo-kernel** [comment]
@nohwnd could you explain which parts you envision "for the rest of the web I want to avoid mixing it with the main repo"

**Comment by nohwnd** [comment]
@bravo-kernel everything except for the docs. Say some main page, articles, talks, list of resources, etc. I'd like to allow others to add their content via PRs to the web, but not mix those PRs that are related to the Pester code+docs, if that makes sense?

**Comment by nohwnd** [comment]
> What will be more present on people's computers nowadays? Ruby or nodejs

I agree on nodejs, that was actually the reason I used hexo in the first place, so I don't have to install ruby. Nowadays I could run it both pretty comfortably using Docker but I did not have that skill before. So even if we do with jekyll we should provide a simple way to work with it.

**Comment by glennsarti** [comment]
@bravo-kernel 
> the reasoning to "avoid SPAs" does not apply to Docusaurus.

Excellent!!! I've seen many SPAs be very bad.  Glad to see it doesn't suffer the same fate.

**Comment by bravo-kernel** [comment]
@nohwnd thanks. As an end-user I would expect all docs to be available in a single place but I can really feel your point; mixing PRs can make things very noisy, very fast.

Having thought about it a little I think my preferred solution would indeed be creating a single repo and thus with both "your" main page, articles, talks, list of resources AND the Get-Help.

The only challenge, regardless of the website technology eventually chosen, would be making sure updates to the code repostory would also trigger an update to the Get-Help but that seems (ci/cd?) solveable.

**Comment by glennsarti** [comment]
> creating a single repo

+1

> updates to the code also trigger an update to the Get-Help but that seems (ci/cd?) solveable.

+1. Defintely do-able with GH Actions.

**Comment by nohwnd** [comment]
> updates to the code also trigger an update to the Get-Help but that seems (ci/cd?) solveable.

Yeah looks like something a webhook trigger from one repo to another would solve.

**Comment by bravo-kernel** [comment]
Right, Docusaurus could do it but all other solutions as well technically. 

The whole reason why I am pitching docusaurus is the "easy of documenation" (all markdown based, just configure and go) + the zero-technical maintenance, this would all be handled by the docusaurus community. All that would remain for pester is theming + the ci/cd.

@glennsarti did you check this [list with possible features](https://github.com/pester/pester.github.com/issues/8#issue-539859130)? We might have missed some there

> HTH, going in await-mode for now

**Comment by nohwnd** [comment]
Okay, I've read through this issue and #8 and I think the Docusaurus looks like a good way forward. The docs will be the meat of the web anyway. I like the way it looks out of the box, but if we get some cool design and make it as refined as this [this one](https://componentkit.org/)  that would be even better. 

So I propose that, @bravo-kernel starts with the setup for docusaurus, then we start migrating the help to show up on the web, and start migrating the wiki. It does not have to be complete, just to see that is is possible.  

Then we can split the work of migrating the content until the documentation is completely in the new approach. I am not sure how much work that will be, I just want to point out that the work can be split if needed, and right now only one cmdlet and one wikipage need to be migrated by @bravo-kernel.

On that is done we can start talking about what would be on the main page and start coming up with a design for the whole thing? 

I have two main points about the help that should be addressed. 
- We should be able to track the v4 and v5 documentation separately. V4 will hopefully go to "maintenance" mode soon, and so the v5 will become the main version of the help. I don't think tracking the minor versions of help is worth the effort. (Right now v4 and v5 are developed in the main Pester repo in `master` and `v5.0` branches)
- The wiki should be migrated to the docs as well. Wiki will then be deprecated and pointed to the new docs page. When v5 is nearing release I will most likely make a copy of the whole thing and put it under v4 for future reference and then improve the wiki for v5. 
 
Also, forget about the `docs.pester.dev` that was just an experiment, maybe we can point it to `pester.dev/docs` later, but let's not use that repo right now.

**Comment by nohwnd** [comment]
Okay folks, we went ahead with Docusaurus, and it looks awesome. Just PR and you get [live preview](https://deploy-preview-1--pester-docs.netlify.com/) and everything. So I assume that 90% of the time, contributors will just edit the markdown and won't have to install anything locally, which is always a plus. It also makes it extremely easy for me to maintain this, because I can see the changes online without taking any manual steps. 

We've set it up about the https://github.com/pester/docs repo. The docs.pester.dev site should be gone now (once dns propagates). I will retire this repo as well, and migrate your access over to /docs

Thanks @bravo-kernel !

**Comment by bravo-kernel** [comment]
Also people, please hold back on PRs until I get the basic structure up-and-running here https://github.com/pester/docs/pull/1. 

Once this is merged in there will be plenty of help needed for:

- theming
- ci/cd
- docs migration
- etc
